feat: Next.js Pub/Sub getting-started guide (#3332)
ably/docs: src/pages/docs/getting-started/nextjs.mdx

}
```
</Code>

Update `src/app/page.tsx` to render the component:

<Code>
```react
// src/app/page.tsx
import { ConnectionState } from './ConnectionState';

export default function Home() {
  return (
    <main style={{ fontFamily: 'Arial, sans-serif', padding: '20px', maxWidth: '900px', margin: '0 auto' }}>
      <h1 style={{ marginBottom: '20px', fontSize: '36px', textAlign: 'center' }}>Ably Pub/Sub - Next.js</h1>
      <ConnectionState />
    </main>
  );
}
```
</Code>

Start the development server:

<Code>
```shell
npm run dev
```
</Code>

Open [http://localhost:3000](http://localhost:3000) and you should see `Connection: connected`. You can also inspect the connection event in the [dev console](https://ably.com/accounts/any/apps/any/console) of your app.

## Step 2: Subscribe to a channel and publish a message <a id="step-2"/>

To publish and subscribe to messages on a channel use the `ChannelProvider` component from the Ably Pub/Sub SDK, which scopes child components to a specific channel.

### ChannelProvider <a id="step-2-channel-provider"/>

The `ChannelProvider` must be nested inside the `AblyProvider`. Update `src/app/page.tsx` to include the `ChannelProvider`:

<Code>

> ```react highlight="+2,+4,+12-14"

Reviewer: I think we either do this for all relevant blocks or none at all 🙂

Author: added everywhere

---
title: "Getting started: Pub/Sub in Next.js"
meta_description: "Get started with Ably Pub/Sub in Next.js. Learn how to publish and subscribe to messages, track presence, retrieve message history, and manage realtime connections in a Next.js App Router application."
meta_keywords: "Pub/Sub Next.js, Next.js PubSub, Ably Next.js SDK, realtime messaging Next.js, publish subscribe Next.js, Ably Pub/Sub guide, Next.js realtime communication, Ably tutorial Next.js, Next.js message history, presence API Next.js, Ably Pub/Sub example, realtime Pub/Sub Next.js"
languages:
  - nextjs
---

> This guide will get you started with Ably Pub/Sub in a new Next.js application using the App Router.

Reviewer: ```suggestion
This guide will get you started with Ably Pub/Sub in a new Next.js application.
```

Author: fixed

1. Install the Ably CLI:

<Code>
```shell
npm install -g @ably/cli
```
</Code>

2. Run the following to log in to your Ably account and set the default app and API key:

<Code>
```shell
ably login
```
</Code>

<If loggedIn={false}>
  <Aside data-type='note'>
  The code examples in this guide include a demo API key. If you wish to interact with the Ably CLI and view outputs within your Ably account, ensure that you replace them with your own API key.
  </Aside>
</If>

## Step 1: Connect to Ably <a id="step-1"/>

Clients establish a connection with Ably when they instantiate an SDK instance. This enables them to send and receive messages in realtime across channels.

Open up the [dev console](https://ably.com/accounts/any/apps/any/console) of your first app before you start so that you can see what happens.

### Set up AblyProvider <a id="step-1-ably-provider"/>

The Ably Pub/Sub SDK provides React hooks and context providers that make it easy to use Pub/Sub features in your components.

> Because the Ably Realtime client uses browser APIs such as WebSocket, it cannot run during server-side rendering. Create a new file `src/app/AblyProvider.tsx` that initializes the client inside a `useEffect` and wraps children in the `AblyProvider`:

Reviewer: Same comment as the others on 'Ably realtime client' vs Pub/Sub SDK.

Author: fixed

<Code>

> ```javascript

Reviewer: This should be React throughout.

Author: fixed

// src/app/AblyProvider.tsx
'use client';

import * as Ably from 'ably';
import { AblyProvider as AblyReactProvider } from 'ably/react';
import { ReactNode, useEffect, useState } from 'react';

export function AblyProvider({ children }: { children: ReactNode }) {
  const [client, setClient] = useState<Ably.Realtime | null>(null);

  useEffect(() => {
    const ably = new Ably.Realtime({
      key: '{{API_KEY}}',
      clientId: 'my-first-client',
    });
    setClient(ably);
    return () => {
      ably.close();
    };
  }, []);

  if (!client) return null;
  return <AblyReactProvider client={client}>{children}</AblyReactProvider>;
}
```
</Code>

<Aside data-type='note'>
Creating the client inside `useEffect` ensures it only runs in the browser, avoiding SSR hydration mismatches. Returning `null` until the client is ready prevents child components from rendering before the Ably context is available. The cleanup function closes the connection when the component unmounts.
</Aside>

Add the `AblyProvider` to your root layout in `src/app/layout.tsx`:

<Code>
```javascript
// src/app/layout.tsx
import { AblyProvider } from './AblyProvider';

> export default function RootLayout({ children }: { children: React.ReactNode }) {

Author: fixed

  useConnectionStateListener((stateChange) => {
    setConnectionState(stateChange.current);
  });

  return (
    <p style={{ color: '#555', padding: '8px 10px', background: '#f0f0f0', borderRadius: '5px' }}>
      Connection: <strong>{connectionState}</strong>
    </p>
  );
}
```
</Code>

Update `src/app/page.tsx` to render the component:

<Code>
```javascript
// src/app/page.tsx
import { ConnectionState } from './ConnectionState';

export default function Home() {
  return (
    <main style={{ fontFamily: 'Arial, sans-serif', padding: '20px', maxWidth: '900px', margin: '0 auto' }}>
      <h1 style={{ marginBottom: '20px', fontSize: '36px', textAlign: 'center' }}>Ably Pub/Sub - Next.js</h1>
      <ConnectionState />
    </main>
  );
}
```
</Code>

Start the development server:

<Code>
```shell
npm run dev
```
</Code>

Open [http://localhost:3000](http://localhost:3000) and you should see `Connection: connected`. You can also inspect the connection event in the [dev console](https://ably.com/accounts/any/apps/any/console) of your app.

## Step 2: Subscribe to a channel and publish a message <a id="step-2"/>

> Messages contain the data that a client is communicating. Channels are named message streams — clients only receive messages published to channels they are subscribed to.

Reviewer: `Channels are named message streams` I'm not sure about this terminology.

Author: changed to Steven's proposal (see his comments)

> Messages contain the data that a client is communicating. Channels are named message streams — clients only receive messages published to channels they are subscribed to.

Reviewer 2: I don't really see the utility in this statement, perhaps it's enough to say.. You can use the ChannelProvider to access the underlying channel, this enables you to publish and subscribe to messages on that channel. The the sections below just detail this?

Author: fixed

### ChannelProvider <a id="step-2-channel-provider"/>

Use the `ChannelProvider` component from the Ably Pub/Sub SDK to scope child components to a specific channel. It must be nested inside the `AblyProvider`.

Update `src/app/page.tsx` to include the `ChannelProvider`:

<Code>

> ```javascript

Reviewer: You could use the syntax highlighting of code blocks for items where we're adding lines in.

Author: added, thanks

        {messages.map((msg) => (
          <p key={msg.id} style={{ margin: '4px 0', borderLeft: '3px solid #007bff', paddingLeft: '8px', color: '#171717' }}>
            <span style={{ color: '#777', marginRight: '6px' }}>{msg.clientId}:</span>
            {String(msg.data)}
          </p>
        ))}
      </div>
      <div style={{ display: 'flex', gap: '8px' }}>
        <input
          type='text'
          value={inputValue}
          placeholder='Type a message...'
          onChange={(e) => setInputValue(e.target.value)}
          onKeyDown={(e) => e.key === 'Enter' && handlePublish()}
          style={{ flex: 1, padding: '10px 12px', border: '1px solid #ccc', borderRadius: '4px', outline: 'none' }}
        />
        <button
          onClick={handlePublish}
          style={{ padding: '10px 20px', background: '#007bff', color: '#fff', border: 'none', borderRadius: '4px', cursor: 'pointer' }}
        >
          Publish
        </button>
      </div>
    </div>
  );
}
```
</Code>

Type a message and click **Publish** to see it appear in your UI. Open another browser window or publish from the CLI to see messages arriving in realtime:

<Code>
```shell

> ably channels publish my-first-channel 'Hello from CLI!'

Reviewer 2: Would it be better to mention subscribe here instead, since you're testing publishing via the component now?

Author: Not sure why it's here, removed.

          </li>
        ))}
      </ul>
    </div>
  );
}
```
</Code>

Update `src/app/page.tsx` to include `PresenceStatus` and `ConnectionState` inside the `ChannelProvider`, alongside `Messages`:

<Code>
```javascript
// src/app/page.tsx
'use client';

import { ChannelProvider } from 'ably/react';
import { ConnectionState } from './ConnectionState';
import { Messages } from './Messages';
import { PresenceStatus } from './PresenceStatus';

export default function Home() {
  return (
    <main style={{ fontFamily: 'Arial, sans-serif', padding: '20px', maxWidth: '900px', margin: '0 auto' }}>
      <h1 style={{ marginBottom: '20px', fontSize: '36px', textAlign: 'center' }}>Ably Pub/Sub - Next.js</h1>
      <ChannelProvider channelName='my-first-channel'>
        <div style={{ display: 'flex', gap: '10px' }}>
          <div style={{ flex: '0 0 220px', display: 'flex', flexDirection: 'column', gap: '10px' }}>
            <PresenceStatus />
            <ConnectionState />
          </div>
          <div style={{ flex: 1 }}>
            <Messages />
          </div>
        </div>
      </ChannelProvider>
    </main>
  );
}
```
</Code>

> Your client ID will appear in the presence list. Test with another client joining via the CLI:

Reviewer: ```suggestion
Your client ID will appear in the presence list. Join presence via the CLI to see another client joining:
```

Author: fixed

<Code>
```shell
ably channels presence enter my-first-channel --data '{"status":"From CLI"}'
```
</Code>

## Step 4: Retrieve message history <a id="step-4"/>

Ably stores messages for 2 minutes by default. You can [extend the storage period](/docs/storage-history/storage) if required.

The `useChannel()` hook returns a `channel` instance. Use its `history()` method to load previously published messages on mount. Update your `Messages` component in `src/app/Messages.tsx` to load history with a `useEffect`:

<Code>
```javascript
// src/app/Messages.tsx
'use client';

import type { Message } from 'ably';
import { useChannel } from 'ably/react';
import { useEffect, useState } from 'react';

export function Messages() {
  const [messages, setMessages] = useState<Message[]>([]);
  const [inputValue, setInputValue] = useState('');

  const { publish, channel } = useChannel('my-first-channel', (message) => {
    setMessages((prev) => [...prev, message]);
  });

  useEffect(() => {
    async function loadHistory() {
      const history = await channel.history({ limit: 5 });
      setMessages(history.items.reverse());
    }
    loadHistory().catch(console.error);

>   }, [channel]);

Author: fixed

            <span style={{ color: '#777', marginRight: '6px' }}>{msg.clientId}:</span>
            {String(msg.data)}
          </p>
        ))}
      </div>
      <div style={{ display: 'flex', gap: '8px' }}>
        <input
          type='text'
          value={inputValue}
          placeholder='Type a message...'
          onChange={(e) => setInputValue(e.target.value)}
          onKeyDown={(e) => e.key === 'Enter' && handlePublish()}
          style={{ flex: 1, padding: '10px 12px', border: '1px solid #ccc', borderRadius: '4px', outline: 'none' }}
        />
        <button
          onClick={handlePublish}
          style={{ padding: '10px 20px', background: '#007bff', color: '#fff', border: 'none', borderRadius: '4px', cursor: 'pointer' }}
        >
          Publish
        </button>
      </div>
    </div>
  );
}
```
</Code>

Publish a few messages first if needed:

<Code>
```shell
ably channels publish --count 5 my-first-channel "Message number {{.Count}}"
```
</Code>

Reload the page. The last 5 messages will appear immediately, loaded from history before any new realtime messages arrive.

### Update globals.css <a id="step-4-globals-css"/>

> Replace the contents of `src/app/globals.css` with the following to reset browser defaults and ensure consistent font sizing across all elements including inputs and buttons:

Reviewer 2: Should this be at the end or at the start? Considering it affects the look of the app as you build it, perhaps starting from the correct style state is better

Author: Moved to prereq.

<Code>
```css
/* src/app/globals.css */
html {
  height: 100%;
}

html,
body {
  max-width: 100vw;
  overflow-x: hidden;
}

body {
  min-height: 100%;
  display: flex;
  flex-direction: column;
  color: #171717;
  background: #ffffff;
  font-family: Arial, Helvetica, sans-serif;
  font-size: 15px;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

*,
input,
button {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
  font-size: inherit;
  font-family: inherit;
}
```
</Code>

Your completed application should look like this:

> ![Screenshot of the completed Next.js Pub/Sub application showing a web interface with a presence panel listing online clients, a connection status indicator, a realtime message feed with blue left-border accents, and a message input field with a Publish button.](../../../images/content/screenshots/getting-started/pub-sub-nextjs-getting-started-guide.png)

Reviewer: That's a very long description of the image :-)

Author: massive one) fixed
